adam-core/src/main/scala/org/bdgenomics/adam/hbase/HBaseFunctions.scala

```diff
@@ -0,0 +1,525 @@
+
+    samples.foreach((x) => {
+      val baos: java.io.ByteArrayOutputStream = new ByteArrayOutputStream()
+      val encoder = EncoderFactory.get().binaryEncoder(baos, null)
+
+      val sampleDatumWriter: DatumWriter[Sample] =
+        new SpecificDatumWriter[Sample](scala.reflect.classTag[Sample].runtimeClass.asInstanceOf[Class[Sample]])
+      sampleDatumWriter.write(x, encoder)
+      encoder.flush()
+      baos.flush()
+
+      val curr_sampleid = x.getSampleId
+
+      val put = new Put(Bytes.toBytes(curr_sampleid))
+      put.addColumn(Bytes.toBytes("meta"), Bytes.toBytes("sampledata"), baos.toByteArray)
+      table.put(put)
+
+    })
+  }
+
+  def loadSampleMetadataFromHBase(hbaseTableName: String,
+                                  sampleIds: List[String]): Seq[Sample] = {
+
+    val conf = HBaseConfiguration.create()
+    val connection = ConnectionFactory.createConnection(conf)
+    val table = connection.getTable(TableName.valueOf(hbaseTableName))
+
+    var resultList = new ListBuffer[Sample]
+
+    sampleIds.foreach((sampleId) => {
+      val myGet = new Get(Bytes.toBytes(sampleId))
+      val result = table.get(myGet)
+      val dataBytes = result.getValue(Bytes.toBytes("meta"), Bytes.toBytes("sampledata"))
+
+      val sampleDatumReader: DatumReader[Sample] =
+        new SpecificDatumReader[Sample](scala.reflect.classTag[Sample]
+          .runtimeClass
+          .asInstanceOf[Class[Sample]])
+
+      val decoder = DecoderFactory.get().binaryDecoder(dataBytes, null)
+
+      while (!decoder.isEnd) {
+        resultList += sampleDatumReader.read(null, decoder)
+      }
+
+    })
+
+    resultList
+
+  }
+
+  def loadRDDofGenotypeFromHBase(sc: SparkContext,
+                                 hbaseTableName: String,
+                                 sampleIds: List[String],
+                                 queryRegion: ReferenceRegion = null,
+                                 numPartitions: Int = 0): RDD[Genotype] = {
+
+    val scan = new Scan()
+    scan.setCaching(100)
+    scan.setMaxVersions(1)
+
+    if (queryRegion != null) {
+      val start = queryRegion.referenceName + "_" + String.format("%10s", queryRegion.start.toString).replace(' ', '0')
+      val stop = queryRegion.referenceName + "_" + String.format("%10s", queryRegion.end.toString).replace(' ', '0')
+      scan.setStartRow(Bytes.toBytes(start))
+      scan.setStopRow(Bytes.toBytes(stop))
+    }
+
+    val conf = HBaseConfiguration.create()
+    val hbaseContext = new HBaseContext(sc, conf)
+
+    sampleIds.foreach(sampleId => {
+      scan.addColumn(Bytes.toBytes("g"), Bytes.toBytes(sampleId))
+    })
+
+    val resultHBaseRDD = hbaseContext.hbaseRDD(TableName.valueOf(hbaseTableName), scan)
+
+    val resultHBaseRDDrepar = if (numPartitions > 0) resultHBaseRDD.repartition(numPartitions)
+    else resultHBaseRDD
+
+    val result: RDD[Genotype] = resultHBaseRDDrepar.mapPartitions((iterator) => {
```

Comment:
Its the genotypes data read in from HBase.
The re partitioning here will have the effect of doing a shuffle prior to deserializing the raw bytes that come back as values from HBase.  Motivation to do so is to increase parallelism at the earliest point possible when, as seems to be the case, bottleneck is not reading data from HBase, but in deserializing it and making it into an ADAM RDD.